### ISSUE TITLE
Using ONNX model for inference

### ISSUE THREAD

**Issue by soltkreig** [open-mmlab/mmpose#949]
Hi! Is it possible to use ONNX model to inference?  Model have a heatmap output and I don't know how to integrate it to inference using mmpose post processing functions

**Comment by wardaddy24** [comment]
Any update? I ported the mmpose model to onnx using the pytorch2onnx.py . I generated the graph from netron.app. The input label is input.1 and needs an image (1,3,256,192). The output layer is labelled as 996. 
On running inference through the 
**config :** mmpose/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/mpii_trb/res101_mpii_trb_256x256.py
**checkpoint(ported to onnx) :** mmpose/checkpoints/res101_mpii_trb_256x256-cfad2f05_20200812.pth
I am getting an output ndarray of size (1, 40, 64, 48).

Can anyone help me with the post-processing of this <class 'numpy.ndarray'>.

**Comment by lunzueta** [comment]
I just wanted the same thing. In the code, you'll see this function: keypoints_from_heatmaps(). it is for that. I hope it helps...

**Comment by ukcastle** [comment]
Hi. I had the same problem
I just tweaked the code and didn't optimize it, but it works.

I used this model
```
from mmpose.apis import init_pose_model
CONFIG = "configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w32_coco_256x192.py"
WEIGHT = "https://download.openmmlab.com/mmpose/top_down/hrnet/hrnet_w32_coco_256x192-c78dce93_20200708.pth"
model = init_pose_model(CONFIG, WEIGHT, "cpu")
```

Post Processing
```
imageHegiht,imageWidth = img.shape[2:]
aspect_ratio = imageWidth / imageHegiht

boxX, boxY, boxW, boxH = 0, 0, 192, 256 #  Cropped Img by RoI
center = torch.tensor([boxX + boxW/2 , boxY + boxH/2])

if boxW > aspect_ratio * boxH:
  boxH = boxW * 1.0 / aspect_ratio
elif boxW < aspect_ratio * boxH:
  boxW = boxH * aspect_ratio

PIXEL_STD = 200
PADDING = 1

heatmaps = output.detach().clone()

detectNum, keypointNum ,heatmapHeigth, heatmapWidth = heatmaps.shape
reshapeHeatmap = torch.reshape(heatmaps, (detectNum,keypointNum,-1))
heatmapMaxIdx = torch.reshape(torch.argmax(reshapeHeatmap, 2),(detectNum,keypointNum,1))
maxVals = torch.reshape(torch.amax(reshapeHeatmap, 2), (detectNum,keypointNum,1))

preds = torch.tile(heatmapMaxIdx, (detectNum,1,2))
preds[:, :, 0] = preds[:, :, 0] % heatmapWidth
preds[:, :, 1] = preds[:, :, 1] // heatmapWidth
preds = torch.where(torch.tile(maxVals, (detectNum,1,2)) > 0.0, preds, -1)
preds = preds.to(torch.float32)

for n in range(detectNum):
  for k in range(keypointNum):
    heatmap = heatmaps[n][k]
    px = int(preds[n][k][0])
    py = int(preds[n][k][1])
    if 1 < px < heatmapWidth-1 and 1 < py < heatmapHeigth-1:
      diff = torch.tensor([
        heatmap[py][px + 1] - heatmap[py][px -1],
        heatmap[py+1][px] - heatmap[py-1][px]
      ])
      preds[n][k] += (torch.sign(diff) * torch.tensor(0.25)) 
coordinate = preds[0]

scale = scale * PIXEL_STD
scaleX = scale[0] / heatmapWidth
scaleY = scale[1] / heatmapHeigth
targetCoordinate = torch.ones_like(coordinate)
targetCoordinate[:, 0] = coordinate[:, 0] * scaleX + center[0] - scale[0] * 0.5
targetCoordinate[:, 1] = coordinate[:, 1] * scaleX + center[1] - scale[1] * 0.5
refineOutput = torch.unsqueeze(torch.cat((targetCoordinate, maxVals[0]),1), 0)

# refineOutput.shape = (1,17,3) , 3 =x,y,pred
```

**Comment by Actmaiji** [comment]
hello, i have some quetions about the post processing.1. what's the different between imageHegiht,imageWidth and boxH, and BoxW?  2.Is the refineOutput coordinate about original image(for eg 441*441 ) or resized image(for eg 192 256)?3.top down(lite-hrnet 18 192*256) processing just directly  resize the image?or use affine transform? Thanks Sincerely for any answer

**Comment by ukcastle** [comment]
@Actmaiji 
Hello, sorry for the late reply because I didn't receive any notice.

When I first made a simple example, I extracted the Bbox roi (Using Human Detection Model) and made a 192x256 image. 
Using that as an example, the xywh format of Bbox is 0,0,192,256. 

It was implemented by following the example in MMPose, so I was trying to find the overall image aspect ratio. 
But when I tried it with onnx, it was most neat to just extract the ROI with the detection model and put it as (1,3,192,256) input.

If it is a 441x441 image and the ROI of a person is (x1:50, y1:50, x2:100, y2:100), Extracting only the ROI (50x50 image) 
I converted it to letterbox format and used it (it didn't come out well if I just resize it)

It's implemented in Yolo v5, so I'll post a link to what I implemented separately.


- [C#](https://github.com/ukcastle/csharp_onnx/blob/5749d23bbb9e0d1ee6a744fe31be535048198e28/onnx_test/OnnxModel.cs#L44)
- [Python](https://github.com/ukcastle/keypoint_pseudo_labeller_mmpose/blob/c6ac732c55b379c0762fcfd6721acdb72fcef73a/src/model_helper.py#L55)
